torch/csrc/distributed/rpc/request_callback_impl.cpp

```diff
@@ -212,14 +212,28 @@ std::shared_ptr<FutureMessage> RequestCallbackImpl::processRpc(
 
       // Process the original RPC.
       auto wrappedMessageType = rpcWithAutograd.wrappedMessageType();
-      auto wrappedRpcResponse = processRpc(
+      // Kick off processing for the nested future and get a Future<T> to the
+      // result.
+      auto wrappedRpcResponseFuture = processRpc(
           rpcWithAutograd.wrappedRpc(), wrappedMessageType, messageId);
-      wrappedRpcResponse->waitNoThrow(); // TODO: make async
 
-      return wrap(getMessageWithAutograd(
-          rpcWithAutograd.fromWorkerId(),
-          std::move(*wrappedRpcResponse).moveValue(),
-          MessageType::FORWARD_AUTOGRAD_RESP));
+      // Make an overall future for the wrapped response.
+      auto responseFuture = std::make_shared<rpc::FutureMessage>();
+      auto fromWorkerId = rpcWithAutograd.fromWorkerId();
+      // The original future needs to be marked as completed when the wrapped
+      // one completes, with the autograd context information wrapped.
+      wrappedRpcResponseFuture->addCallback(
+          [responseFuture, messageId, fromWorkerId, wrappedRpcResponseFuture](
+              const Message& m,
+              const c10::optional<utils::FutureError>& error) {
+            // Hack since getMessageWithAutograd requires a Message&&
+            auto msg = m;
```

Comment:
Instead of forcing a copy here, you can probably do something like  wrappedRpcResponseFuture->moveValue()

Also, I think you need to propagate any error